Google.Api.Generator.Tests/WhitespaceFormatterTest.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This test loads this source file during the test, parses it into
+// Roslyn, removes all trivia, passes it through the Whitespace formatter,
+// then checks that the input source and re-formatted source are identical.
+// Code at the start, end, and between IGNORE_START/IGNORE_END markers
+// are ignored.
+// So some of the source code looks a bit odd, as it needs to test every
+// code construct is formatted correctly.
+// TODO: Add further code as WhitespaceFormatter adds more formatting.
+
+// IGNORE_END
+using Google.Api.Generator.Formatting;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Xunit;
+using sys = System;
+
+namespace Google.Api.Generator.Tests
```

Comment:
Is there a significant benefit in making the test's own source code (with bits ignored) the test? While it's amusing/elegant, it feels like it may be making things trickier than expected.

Reply:
I think keeping this test within a single source file is useful.
I've added a TODO to reconsider this if splitting the test would be more comprehensible.